### ISSUE TITLE
Support for directives

### ISSUE THREAD

**Issue by davej** [angular/router#161]
If the router supported ['component-based'](https://www.airpair.com/angularjs/posts/component-based-angularjs-directives) directives (including the directive's template) then it would give a nice path for migrating controllers/components to angular 2.0.

Any thoughts on achieving something like below?:

```javascript
angular.module('app', ['ngNewRouter'])
  .directive('componentDirective', function() {
    return {
      templateUrl : 'componentDirective.html',
      controllerAs : 'ctrl',
      controller : function() {
        // Controller here
      }
    }
  });
  .controller('AppController', function($router) {
    $router.config([
      {path: '/', component: 'componentDirective' }
    );
  });
```

**Comment by hannahhoward** [comment]
+1 a directive seems more analogous to a component than a controller + view. on the other hand it's been pointed out several times that the 1.x DDO api is very complicated for beginners.

**Comment by threehams** [comment]
+1 Been using this pattern for a while, rather than the module approach in the examples. Changing to a component structure through modules would be a pretty major change, which I wouldn't want to make unless it's part of a more official 2.0 migration path.

Maybe allow {path: '/', directive: 'componentDirective'} for this approach? I'm not a fan of complicating APIs with extra options, but it would be more explicit about intention.

**Comment by btford** [comment]
Thanks for the proposal, @davej! I'm definitely considering it.

**Comment by btford** [comment]
See also – https://github.com/angular/angular.js/issues/10007

**Comment by davej** [comment]
Yeah, when I heard the identifier for the router was 'component', I assumed that it was a tie-in with the new component helper syntax. I was surprised when the new router didn't support it. Anyway, great news! :-)

Component directives + the new router is a good bridge to making the 2.0 transition easier.

**Comment by jeffwhelpley** [comment]
This is a no brainer. Implement angular.component and then use that in the router. I wouldn't bother with the directive implementation unless for some reason you don't want to build angular.component in 1.5.

**Comment by davej** [comment]
>  I wouldn't bother with the directive implementation unless for some reason you don't want to build angular.component in 1.5.

`angular.component` is just a shorthand for a directive, the router does not need to make a distinction between a component or a directive, nor should it IMO.

**Comment by jeffwhelpley** [comment]
Sure, it's no big deal either way. I was just saying don't bother because the point of angular.component is to align your app with 2.0. There are no current apps using the new router since it just came out so you might as well just have everyone move right do components. But, sure, if you want to use directives for some reason.

**Comment by timkindberg** [comment]
:+1: I thought of this as soon as I heard the new router was component based as did several of my coworkers. In fact I've heard one or two people say it's "directive-based", clearly misunderstanding, but justified I think. Some way to load a directive that has a template and controller just makes sense.

**Comment by busypeoples** [comment]
+1 This is what I thought the new router would be able to do out of the box. Would be very helpful if there was the option to use directives as mentioned by @threehams.

> angular.component is just a shorthand for a directive, the router does not need to make a distinction between a component or a directive, nor should it IMO.

@davej +1

**Comment by butchpeters** [comment]
How would this work with a directive (component) that has attributes?  Normally, I would consume a directive like a web component might be consumed:
```html
<my-component attr1="value1" attr2="value2"></my-component>
```
So, if I want to use the router to compose my UI now, I would replace the directive with a viewport
```html
<div ng-viewport="left"></div>
```
and then load my directive with the route config?
```javascript
$router.config([
  {path: '/somepath', component: {left: 'myComponent', right: 'somethingElse'}}
]);
```
Would I need to somehow configure the $routeParams to contain all the attributes that all my components will need, or will there be another way to do this?

**Comment by davej** [comment]
How about an API like this for using directive attributes instead of route params.
Here's a quick example of an article component for a blog.

```javascript
angular.module('blog', ['ngNewRouter'])
  .directive('articleComponent', function() {
    return {
      templateUrl : 'article.html',
      controllerAs : 'ctrl',
      bindToController: {
        articleId: '@'
      },
      controller: function($http) {
        $http.get('http://myblog.com/articles/' + this.articleId + '.json')
          .then(function(response) {
            this.currentArticle = response.data;
          }.bind(this));
      }
    };
  })
  .controller('BlogController', function($router) {
    // begin router
    $router.config([
      { path: 'article/:articleId', component: 'articleComponent' }
    ]);
    // end router
  });
```

With this API, a route like this:
```
GET /article/34
```

Would be equivalent to the following component/directive in HTML:
```HTML
<article-component article-id="34"></article-component>
```

This feels like a super clean cross-over to me. Thoughts?

**Comment by timkindberg** [comment]
@davej that actually seems pretty intuitive! 

We might still need a way to pass additional params though. That would be inline with the `params` config that was proposed in this issue: https://github.com/angular/router/issues/121#issuecomment-76597328.

**Comment by busypeoples** [comment]
@timkindberg From what I understand, you can't pass any additional params in the current implementation either:
```javascript
function DetailController ($routeParams) {
  this.id = $routeParams.id;
}
```
Please correct if mistaken here.
I could imagine that the directive might retrieve any further information or initiate the component state via a service or load any data via a store. 
I think the outlined approach by @davej seems very straight forward.

**Comment by ChristianKohler** [comment]
How about aligning the Angular components (aka. simple controllers) and Angular directives. The way how we use controllers with the new router is basically a component without access to the element. 

We could just use directives as components. We wouldn't resolve the template urls by convention.

```js
.directive('home', function() {
    return {
      templateUrl : 'home.html',
      controllerAs : 'home',
      controller: function() {
        this.welcome = "Hi";
      }
    };
  })
```

When we need more control and acces to the element, we would just use the link or compile function. This approach would play nicely with https://github.com/angular/angular.js/issues/10007

**Comment by hannahhoward** [comment]
@btford are you already implementing this? I was thinking of giving this a try over the weekend but if there is work already in progress I'll hold off.

**Comment by hrajchert** [comment]
+1 I like @davej way of passing parameters.

And I think it aligns with proposals in https://github.com/angular/angular.js/issues/10007 about how to pass parameters.

**Comment by hannahhoward** [comment]
I submitted a pull request to implement this feature as specified several weeks ago. No feedback from the core contributors or merge unfortunately. Not sure what else to do or if there was something missing from the PR I submitted.

**Comment by eXaminator** [comment]
Would anyone care to join the discussions on the PR? There are some problems discussed there and some more input might help to solve those.

**Comment by mischkl** [comment]
@btford bump. We are building our code base to be Angular 2.0-compatible and easily reusable by packaging our components as Directives + Controllers. Now I have no idea how to integrate this way of working with the new Component Router....

**Comment by PascalPrecht** [comment]
@btford is currently busy pushing the Angular 2 implementation forward. APIs will be streamlined afterwards I guess. And there's nothing stopping us from already putting everything in component directives and wrap them with a controller/template to make them usable today. Migration will be easy once APIs have been stabilized.

**Comment by mischkl** [comment]
@PascalPrecht makes sense. Since this is an Angular 1.x-specific request I guess the best thing to do is wait for the Angular 2 implementation to be stabilized first. But of course since our current need is for the 1.x implementation it feels more important to me. ;)
